speculos/client.py

```diff
@@ -0,0 +1,193 @@
+
+    with Image.open(path1) as img1:
+        with Image.open(path2) as img2:
+            diff_img = ImageChops.difference(img1, img2)
+    return diff_img.getbbox() is None
+
+
+class Api:
+    def __init__(self, api_url: str) -> None:
+        self.api_url = api_url
+        self.timeout = 2000
+        self.session = requests.Session()
+        self.stream = self._open_stream()
+
+    def _open_stream(self) -> requests.Response:
+        stream = self.session.get(f"{self.api_url}/events?stream=true", stream=True)
+        check_status_code(stream, "/events")
+        return stream
+
+    def get_next_event(self) -> dict:
+        line = self.stream.raw.readline()
+        event = json.loads(line)
+        if not isinstance(event, dict):
+            raise ClientException(f"Invalid event ({event})")
+        return event
+
+    def wait_for_text_event(self, text: str) -> dict:
+        """Wait until an event containing the specified text is received."""
+
+        while True:
+            event = self.get_next_event()
+            if text in event["text"]:
+                break
+        return event
+
+    def press_and_release(self, button: str) -> None:
+        assert button in ["left", "right", "both"]
+        data = {"action": "press-and-release"}
+        with self.session.post(f"{self.api_url}/button/{button}", json=data) as response:
+            check_status_code(response, f"/button/{button}")
+
+    def finger_touch(self, x: int, y: int) -> None:
+        data = {"action": "press-and-release", "x": x, "y": y}
+        with self.session.post(f"{self.api_url}/finger", json=data) as response:
+            check_status_code(response, "/finger")
+
+    def get_screenshot(self) -> bytes:
+        with self.session.get(f"{self.api_url}/screenshot") as response:
+            check_status_code(response, "/screenshot")
+            return response.content
+
+    def _apdu_exchange(self, data: bytes) -> bytes:
+        with self.session.post(f"{self.api_url}/apdu", json={"data": data.hex()}) as response:
+            apdu_response = ApduResponse(response)
+            return apdu_response.receive()
+
+    def _apdu_exchange_nowait(self, data: bytes) -> requests.Response:
+        return self.session.post(f"{self.api_url}/apdu", json={"data": data.hex()}, stream=True)
+
+    def set_automation_rules(self, rules: dict) -> None:
+        with self.session.post(f"{self.api_url}/automation", json=rules) as response:
+            check_status_code(response, "/automation")
+
+
+class SpeculosInstance:
+    def __init__(self, app: str, args: List[str] = []) -> None:
+        self.app = app
+        self.args = args
+        self.process: Optional[subprocess.Popen] = None
+
+        if "--display" not in self.args:
+            self.args += ["--display", "headless"]
+
+        if "--api-port" not in self.args:
+            self.port = 5000
+        else:
+            n = self.args.index("--api-port")
+            self.port = int(self.args[n])
```

Comment:
This function has two bugs:

1. Using `args=[]` as parameter has the unexpected side effet of declaring a static list associated with the function. So every time `__init__` is called without specifying `args`, parameters are *added* to `args`. Moreover if a caller calls the constructor with code such as:
```python
my_args = ["--apt-port", "5000"]
instance1 = SpeculosInstance("app", my_args)
my_args[1] = "5001"
instance2 = SpeculosInstance("app", my_args)
```
... in fact `my_args[1] = "5001"` will modify `instance1.args` unexpectedly. I suggest modifying the code to:

```python
from typing import Iterable
# ...
class SpeculosInstance:
    def __init__(self, app: str, args: Iterable[str] = ()) -> None:
        self.app = app
        self.args = list(args)
# ...
```

2. `self.port = int(self.args[n])` calls `int` on `"--api-port"` instead of the API port value.  `self.port = int(self.args[n + 1])` should be used instead.

Reply:
Nice catch!